### ISSUE TITLE
Shibboleth: permit use of request headers rather than environment variables

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#2294]
#2180 is about reintroducing Apache for Shibboleth support but we've discovered that the default and recommended setup requires AJP which causes problems with certain file downloads as explained in detail at https://github.com/payara/Payara/issues/350 and https://java.net/jira/browse/GRIZZLY-1787

For security reasons, the Shibboleth wiki strongly recommends using environment variables via the AJP protocol but mentions that "data can be passed via HTTP request headers (by means of using ShibUseHeaders On) but this is considered less secure" at https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPJavaInstall

This ticket is about making the Dataverse code more flexible to permit this less-secure "headers" mode via a configuration option so that we can avoid using AJP. That is to say, if ShibUseHeaders is set to true the application will use `request.getHeader()` rather than `request.getAttribute()`.

In order for the application to start receiving the headers, however, the `Location` stanza documented at http://guides.dataverse.org/en/4.0.1/installation/shibboleth.html#edit-apache-config-files needs to include an additional "ShibUseHeaders On" directive like this:

```
<Location /shib.xhtml>
  AuthType shibboleth
  ShibRequestSetting requireSession 1
  require valid-user
  ShibUseHeaders On
</Location>
```

To switch from AJP to HTTP, in the same file, the line...

```
ProxyPass / ajp://localhost:8009/
```

... would be changed to this:

```
ProxyPass / http://localhost:8080/
```

This is the difference between using `mod_proxy_ajp` and `mod_proxy_http`, both of which are modules that ship with Apache. Apache needs to be restarted to pick up the change.

Here's a curl command to enable the ShibUseHeaders option:

`curl -X PUT -d true http://localhost:8080/api/admin/settings/:ShibUseHeaders`

It can't be overstated how strongly the Shibboleth wiki recommends against using headers. For example:

> "Use of this option [ShibUseHeaders] should be avoided." https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPApacheConfig

> "Under no circumstances should you rely on the request header option other than as a temporary measure while adjusting applications to use the environment option. There are no known scenarios in which environment variables can't be used, including with Java containers, though sometimes extra effort or Apache settings may be needed. Do NOT take shortcuts with this. Do the work and use them." https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPSpoofChecking

**Comment by pdurbin** [comment]
Implemented in b64623c. Passing to QA.

**Comment by pdurbin** [comment]
Obviously, I'm concerned about the security implications of this change so I ran this issue and the commit above past the Shibboleth mailing list:

AJP, headers, security - http://shibboleth.net/pipermail/users/2015-June/022456.html

**Comment by pdurbin** [comment]
In a [reply by Scott Cantor](http://shibboleth.net/pipermail/users/2015-June/022458.html) some tips were offered on testing the security of the new ShibUseHeaders setting...

On Mon, Jun 29, 2015 at 4:19 PM, Cantor, Scott wrote:
>
> On 6/29/15, 3:34 PM, "users on behalf of Philip Durbin" <users-bounces@shibboleth.net on behalf of philip_durbin@harvard.edu> wrote:
> >
> >1. Am I right in thinking that to avoid using AJP I have to set
> >"ShibUseHeaders" to "On"?
>
> Yes.
>
> >2. Can anyone provide any guidance in how I could try to hack into my
> >application when "ShibUseHeaders On" is enabled?
>
> Trivially, if you can access it directly. Try the [Modify Headers add-on for Firefox](https://addons.mozilla.org/En-us/firefox/addon/modify-headers/). If you can't access it directly, then it is only vulnerable if and when the SP springs a leak or something happens configuration-wise to change which headers the SP is managing that doesn't line up with what the application is reading.
>
> Also depends how the headers are being propagated. Explicit propagation via Apache mod_headers is better/safer than connectors that auto-forward all headers such as WebLogic's module does.

... however, @kcondon and @scolapasta and I met today and we are hopeful the we won't even need this "use headers" option since the "go back to using AJP" option at #2180 seems to be working (still needs to go through QA though).

**Comment by kcondon** [comment]
This approach was sidelined because we found a fix for the ajp approach. Closing.

**Comment by pdurbin** [comment]
While working on #2916 I just realized that it's still possible to configure Dataverse to use request headers. We might want to consider reverting b64623c to disallow this completely. Certainly, I do not plan to document the boolean that controls it while I work on #2953 because I feel so strongly that this configuration should never be used, based on the many stern warnings I've quoted from the Shibboleth docs/wiki in this issue.
